pkg/admission/pod_eviction.go

```diff
@@ -0,0 +1,464 @@
+	errPodNotReady                = "pod is not ready"
+	errNoPdbForPod                = "no zone pod disruption budgets found for pod"
+	errMaxUnavailableIs0          = "max unavailable = 0"
+	errUnableToDetermineStsStatus = "unable to determine pod status in related StatefulSet"
+
+	// the AdmissionResponse HTTP status codes sent in deny responses
+	// PDB has been reached
+	httpStatusPdbExceeded = 429
+	// maxUnavailable=0
+	httpStatusPdbDisabled = 403
+	// unable to determine status of pods/statefulsets
+	httpStatusInternalError = 500
+	// configuration error
+	httpStatusMisconfiguration = 400
+)
+
+// A partitionMatcher is a utility to assist in matching pods to a partition.
+type partitionMatcher struct {
+	same func(pod *corev1.Pod) bool
+}
+
+// A pdbValidator abstracts the different ZPDB implementations
+type pdbValidator struct {
+
+	// considerSts returns true if this StatefulSet should be considered in the PDB tallies
+	considerSts func(sts *appsv1.StatefulSet) bool
+
+	// considerPod returns a matcher which is used to test if a pod should be considered in the PDB tallies
+	considerPod partitionMatcher
+
+	// accumulateResult is called for each StatefulSet which is tested
+	accumulateResult func(sts *appsv1.StatefulSet, result *podStatusResult) error
+
+	// validate is called after all the StatefulSets have been tested - this function validates that the PDB will not be breached
+	validate func(maxUnavailable int) error
+
+	// successMessage returns a success message we return in the eviction response
+	successMessage func() string
+}
+
+// plural appends an 's' to the given string if the value is > 1.
+func plural(s string, value int) string {
+	if value > 1 {
+		return fmt.Sprintf("%ss", s)
+	}
+	return s
+}
+
+// pdbMessage creates a message which includes the number of not ready and unknown pods within the given span
+func pdbMessage(result *podStatusResult, span string) string {
+	msg := ""
+	if result.notReady > 0 {
+		// 1 pod not ready
+		msg += fmt.Sprintf("%d %s not ready", result.notReady, plural("pod", result.notReady))
+		if result.unknown > 0 {
+			msg += ", "
+		}
+	}
+	if result.unknown > 0 {
+		// 2 pods unknown
+		msg += fmt.Sprintf("%d %s unknown", result.unknown, plural("pod", result.unknown))
+	}
+
+	// under ingester-zone-a partition 0
+	// under ingester-zone-a
+	if len(span) > 0 {
+		msg += fmt.Sprintf(" under %s", span)
+	}
+	return msg
+}
+
+// a lock used to control finding a specific PDB lock
+var lock = sync.RWMutex{}
+
+// a map of ZPDB name to lock
+var locks = make(map[string]*sync.Mutex, 5)
+
+// findLock returns a Mutex for each unique name, creating the Mutex if required.
+func findLock(name string) *sync.Mutex {
+	lock.RLock()
+	if pdbLock, exists := locks[name]; exists {
+		lock.RUnlock()
+		return pdbLock
+	}
+	lock.RUnlock()
+
+	lock.Lock()
+	defer lock.Unlock()
+
+	if pdbLock, exists := locks[name]; exists {
+		return pdbLock
+	}
+
+	locks[name] = &sync.Mutex{}
+	return locks[name]
+}
```

Comment:
Rather than creating global variables like this, I would create some kind of struct to encapsulate all the state for the eviction endpoint (eg. `EvictionController` or something like that) that has the `PodEviction` method on it.